src/v/kafka/protocol/schemata/generator.py

```diff
@@ -309,6 +315,33 @@
     fetch_record_set=("batch_reader", None, "read_nullable_batch_reader()"),
 )
 
+
+def snake_case(name):
+    """Convert camel to snake case"""
+    return name[0].lower() + "".join(
+        [f"_{c.lower()}" if c.isupper() else c for c in name[1:]])
+
+
+# Declare some structs as sensitive. Utmost care should be taken to ensure the
+# contents of these structs are never made available over unsecure channels
+# (sent over an unencrypted connection, printed in logs, etc.).
+sensitive_msg_names = [snake_case(f"{msg}Data") for msg in [
```

Comment:
looks good.

i think i was able to get the code generator to obfuscate sensitive data at a finer granularity so we don't need to redact the entire structure. this seems to produce the results we want:

```diff
diff --git a/src/v/kafka/protocol/schemata/generator.py b/src/v/kafka/protocol/schemata/generator.py
index 62aa4e524..cbadef532 100644
--- a/src/v/kafka/protocol/schemata/generator.py
+++ b/src/v/kafka/protocol/schemata/generator.py
@@ -328,6 +328,14 @@ struct_renames = {
 }
 # yapf: enable
 
+sensitive_map = {
+    "SaslAuthenticateRequestData": {
+        "AuthBytes": True,
+    },
+    "SaslAuthenticateResponseData": {
+        "AuthBytes": True,
+    }
+}
 
 def make_context_field(path):
     """
@@ -679,6 +687,18 @@ class Field:
 
         raise Exception(f"No decoder for {(tn, fn)}")
 
+    def _sensitive_path(self):
+        """
+        Resolve a redpanda field path sensitive annotation.
+        """
+        d = sensitive_map
+        for p in self._path:
+            d = d.get(p, None)
+            if d is None:
+                break
+        assert d is None or d is True
+        return d
+
     @property
     def decoder(self):
         """
@@ -706,6 +726,10 @@ class Field:
         assert plain_decoder[1]
         return plain_decoder[1], named_type
 
+    @property
+    def sensitive(self):
+        return self._sensitive_path()
+
     @property
     def is_array(self):
         return isinstance(self._type, ArrayType)
@@ -934,11 +958,11 @@ void {{ struct.name }}::decode(iobuf, api_version) {}
 {% set structs = struct.structs() + [struct] %}
 {% for struct in structs %}
 {%- if struct.fields %}
-std::ostream& operator<<(std::ostream& o, const {{ struct.name }}& v) {
+std::ostream& operator<<(std::ostream& o, [[maybe_unused]] const {{ struct.name }}& v) {
     fmt::print(o,
       "{{'{{' + struct.format + '}}'}}",
       {%- for field in struct.fields %}
-      v.{{ field.name }}{% if not loop.last %},{% endif %}
+      {%- if field.sensitive %}"***"{% else %}v.{{ field.name }}{% endif %}{% if not loop.last %},{% endif %}
       {%- endfor %}
     );
     return o;
```

with this diff i was able to get the following operator<< overload to generate

```c++
std::ostream& operator<<(std::ostream& o, [[maybe_unused]] const sasl_authenticate_response_data& v) {
    fmt::print(o,
      "{{error_code={} error_message={} auth_bytes={} session_lifetime_ms={}}}",v.error_code,v.error_message,"***",v.session_lifetime_ms
    );
    return o;
}
```

Reply:
Nice! I'd be happy to adopt this, though before doing so, I think it's also worth further discussion into what fields _shouldn't_ be redacted. Without deeper knowledge about the messages and how they're used I'm not sure it'll be clear to decide.

I might also be inclined to trudge forward with this coarse grained approach while we decide on the details of the finer grained field-oriented approach. I understand it may hurt debuggability, but I'm not sure how much of a hit it'll be. In supporting deployments so far, do we frequently (or ever) find these stream operators useful?

Reply:
yeh let's have a discussion on slack or wherever is appropriate as soon as possible. redacting entire swaths of responses is safe, but also makes debugging very difficult.

Reply:
This is pretty cool! I took a stab at this and my approach was a new type that would have an overloaded operator<< that prints "***" and have the generator wrap the types we declare as secrets instead. Also is nice because in the generated code you'll see the type as `security::secret<bytes> auth_bytes;` Both seem like good solutions.

Reply:
I've combed through the Kafka schemata and the following seem like fields we should be redacting. Lmk if you think there are any other worth inclusions:
```
AlterConfigsRequest: Value
DescribeConfigsResponse: Value (there is an IsSensitive field here we may be able to use; not sure how important it is for debugging to honor it, vs be aggressive and redact everything)
FetchResponse: Records
IncrementalAlterConfigsRequest: Value
ProduceRequest: Records
SaslAuthenticateRequest: AuthBytes
SaslAuthenticateResponse: AuthBytes
```

Some others I considered but I don't think qualify:
```
CreateTopics: Configs.Value (seems like these are all tuning parameters that affect topic maintenance, rather than any kind of secret)
ACLs: we discussed this briefly in our call. While it's not something we should always be broadcasting, it's probably not a huge deal if this ends up in logs, and it may help with debugging the odd security failure
```

Reply:
>Any reason you're using a string=>bool map instead of a set of strings to declare sensitivity?

Answering this myself, seems this makes the code structure easier when considering the hierarchical nature of the message bodies.

Reply:
> This is pretty cool! I took a stab at this and my approach was a new type that would have an overloaded operator<< that prints "***" and have the generator wrap the types we declare as secrets instead. Also is nice because in the generated code you'll see the type as security::secret<bytes> auth_bytes; Both seem like good solutions.

fixes that take advantage of the cpp type system are preferred if they are ready @graphcareful

Reply:
Opened https://github.com/redpanda-data/redpanda/issues/4984 to track this

Reply:
@dotnwat i have a branch where i introduced this type but i wanted to use the type everywhere instead of just immediately unwrapping it as soon as it is parsed and passed to the request handler. Opinions?